parser/checkpoint.go

```diff
@@ -39,27 +39,141 @@ func (c *CheckpointEvent) Parse(r reader.Reader, classes ClassMap, cpools PoolMa
 		if err != nil {
 			return fmt.Errorf("unable to parse constant pool class: %w", err)
 		}
+		m, err := r.VarInt()
+		if err != nil {
+			return fmt.Errorf("unable to parse constant pool's number of constants: %w", err)
+		}
 		cm, ok := cpools[int(classID)]
 		if !ok {
-			cpools[int(classID)] = &CPool{Pool: make(map[int]ParseResolvable)}
+			cpools[int(classID)] = &CPool{Pool: make(map[int]ParseResolvable, m)}
 			cm = cpools[int(classID)]
 		}
-		m, err := r.VarInt()
-		if err != nil {
-			return fmt.Errorf("unable to parse constant pool's number of constants: %w", err)
+		class, ok := classes[int(classID)]
+		if !ok {
+			return fmt.Errorf("unexpected class %d", classID)
+		}
+		var (
+			results            []ParseResolvable
+			preAllocateResults = true
+			contantsSlice      = getConstantsSlice(int(m), class.numConstants)
+		)
+		// preallocate common class in async-profiler
+		results = make([]ParseResolvable, m)
+		switch class.Name {
+		case "java.lang.Thread":
+			threads := make([]Thread, m)
+			for i := range threads {
+				threads[i].constants = contantsSlice[i]
+				results[i] = &threads[i]
+			}
+		case "jdk.types.StackTrace":
+			var classStackFrames *ClassMetadata
+			for _, class := range classes {
+				if class.Name == "jdk.types.StackFrame" {
+					classStackFrames = class
+					break
+				}
+			}
+			var (
+				stackFrames      []StackFrame
+				indexStackFrames int
+			)
+			createStackFrames := func() ParseResolvable {
+				if indexStackFrames >= len(stackFrames) {
+					stackFrames = make([]StackFrame, m)
+					contantsSlice = getConstantsSlice(int(m), classStackFrames.numConstants)
+					for i := range stackFrames {
+						stackFrames[i].constants = contantsSlice[i]
+					}
+					indexStackFrames = 0
+				}
+				result := &stackFrames[indexStackFrames]
+				indexStackFrames++
+				return result
+			}
+			for i, class := range classes {
+				if class.Name == "jdk.types.StackFrame" {
+					class.typeFn = createStackFrames
+					classes[i] = class
+					break
+				}
+			}
+			var pointerToStackFrames []*StackFrame
+			indexPointerToStackFrames := 0
+			getPointerToStackFrames := func(n int) []*StackFrame {
+				if n > int(m) {
+					pointerToStackFrames = make([]*StackFrame, n)
```

Comment:
I think there's a problem somewhere in this function. Probably in this branch. Maybe It should either adjust `indexPointerToStackFrames` or do not modify `pointerToStackFrames`

Here's an unexpected behavior:
https://go.dev/play/p/y1_U-6bT7YW

We should at least fix the problem

I also recommend moving `getPointerToStackFrames` and `createStackFrames` to a separate types, something like `StackFrameSliceAllocator` and `StackFrameAllocator` and cover it with tests. But that is not a requirement, the bare minimum is to fix the problem

Reply:
I've reopened this commend since I don't see how this is resolved and this is the only concerning question I have.

Reply:
> I've reopened this commend since I don't see how this is resolved and this is the only concerning question I have.

sorry lost it